packages/core/core/src/AssetGraph.js

```diff
@@ -154,42 +158,54 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
         ...assetGroups.map(assetGroup => nodeFromAssetGroup(assetGroup)),
       );
     }
-    this.replaceNodesConnectedTo(nullthrows(this.getRootNode()), nodes);
+    let nodeIds = [];
+    for (let node of nodes) {
+      nodeIds.push(this.addNode(node));
+    }
+    this.replaceNodeIdsConnectedTo(nullthrows(this.rootNodeId), nodeIds);
   }
 
-  addNode(node: AssetGraphNode): AssetGraphNode {
+  addNode(node: AssetGraphNode): NodeId {
     this.hash = null;
-    return super.addNode(node);
+    return super.addNodeByContentKey(node.id, node);
   }
 
-  removeNode(node: AssetGraphNode): void {
+  removeNode(nodeId: NodeId): void {
     this.hash = null;
-    this.onNodeRemoved && this.onNodeRemoved(node);
+    this.onNodeRemoved && this.onNodeRemoved(nodeId);
     // This needs to mark all connected nodes that doesn't become orphaned
     // due to replaceNodesConnectedTo to make sure that the symbols of
     // nodes from which at least one parent was removed are updated.
-    if (this.isOrphanedNode(node) && node.type === 'dependency') {
-      let children = this.getNodesConnectedFrom(node);
+    let node = nullthrows(this.getNode(nodeId));
+    if (this._isOrphanedNode(nodeId) && node.type === 'dependency') {
+      let children = this.getNodeIdsConnectedFrom(nodeId).map(id =>
+        nullthrows(this.getNode(id)),
+      );
       for (let n of children) {
         invariant(n.type === 'asset_group' || n.type === 'asset');
         n.usedSymbolsDownDirty = true;
       }
     }
-    return super.removeNode(node);
+    return super.removeNode(nodeId);
   }
 
   resolveEntry(
     entry: string,
     resolved: Array<Entry>,
-    correspondingRequest: string,
+    correspondingRequest: ContentKey,
   ) {
-    let entrySpecifierNode = nullthrows(
-      this.getNode(nodeFromEntrySpecifier(entry).id),
+    let entrySpecifierNodeId = this.getNodeIdByContentKey(
+      nodeFromEntrySpecifier(entry).id,
     );
+    let entrySpecifierNode = nullthrows(this.getNode(entrySpecifierNodeId));
     invariant(entrySpecifierNode.type === 'entry_specifier');
     entrySpecifierNode.correspondingRequest = correspondingRequest;
-    let entryFileNodes = resolved.map(file => nodeFromEntryFile(file));
-    this.replaceNodesConnectedTo(entrySpecifierNode, entryFileNodes);
+
+    let entryFileNodeIds = [];
+    resolved.map(file => {
+      entryFileNodeIds.push(this.addNode(nodeFromEntryFile(file)));
+    });
+    this.replaceNodeIdsConnectedTo(entrySpecifierNodeId, entryFileNodeIds);
```

Comment:
nit: `replaceNodeIdsConnectedTo(..., resolved.map(node => this.addNode(nodeFromEntryFile(file))); `